src/neuroglancer/datasource/deepzoom/backend.ts

```diff
@@ -0,0 +1,132 @@
+  gridShape = (() => {
+    const gridShape = new Uint32Array(3);
+    const {upperVoxelBound, chunkDataSize} = this.spec;
+    for (let i = 0; i < 3; ++i) {
+      gridShape[i] = Math.ceil(upperVoxelBound[i] / chunkDataSize[i]);
+    }
+    return gridShape;
+  })();
+
+  async download(chunk: VolumeChunk, cancellationToken: CancellationToken): Promise<void> {
+    const {parameters} = this;
+
+    // /* This block is enough if support for these aren't needed:
+    //  * - Firefox before 105 (OffscreenCanvas, 2022-09-20)
+    //  * - Safari before 16.4 (OffscreenCanvas, 2023-03-27)
+    //  */
+    // const {tilesize, overlap} = parameters;
+    // const [x, y] = chunk.chunkGridPosition;
+    // const url = `${parameters.url}/${x}_${y}.${ImageTileEncoding[parameters.encoding].toLowerCase()}`;
+    // const response: Blob = await cancellableFetchSpecialOk(this.credentialsProvider, url, {}, response => response.blob(), cancellationToken);
+    // const tile = await createImageBitmap(response);
+    // const canvas = new OffscreenCanvas(tilesize, tilesize);
+    // const ctx = canvas.getContext("2d")!;
+    // ctx.drawImage(tile, x === 0 ? 0 : -overlap, y === 0 ? 0 : -overlap);
+    // const id = ctx.getImageData(0, 0, tilesize, tilesize).data;
+    // const t2 = tilesize * tilesize;
+    // const d = chunk.data = new Uint8Array(t2 * 3);
+    // for (let i = 0; i < t2; i++) {
+    //   d[i] = id[i * 4];
+    //   d[i + t2] = id[i * 4 + 1];
+    //   d[i + 2 * t2] = id[i * 4 + 2];
+    // }
+    // Todo: ^ "transposeArray2d" likely does the same
+
+    const {tilesize, overlap, encoding} = parameters;
+    const [x, y] = chunk.chunkGridPosition;
+    const ox = x === 0 ? 0 : overlap;
+    const oy = y === 0 ? 0 : overlap;
+    const url = `${parameters.url}/${x}_${y}.${ImageTileEncoding[encoding].toLowerCase()}`;
+    try{
+      const responseBuffer = await cancellableFetchSpecialOk(
+          this.credentialsProvider, url, {}, responseArrayBuffer, cancellationToken);
+
+      const responseDataView = new DataView(responseBuffer);
+      let tilewidth = 0, tileheight = 0;
+      let tiledata: Uint8Array|undefined;
+      switch(encoding){
+        case ImageTileEncoding.PNG:
+          /* Somewhat straightforward, width and height are at fixed position
+           * no validation attempt, decoding throws on mismatch/garbage anyway
+           */
+          tilewidth = responseDataView.getInt32(16, false);
+          tileheight = responseDataView.getInt32(20, false);
+          tiledata = await requestAsyncComputation(
+              decodePng, cancellationToken, [responseBuffer],
+              new Uint8Array(responseBuffer), tilewidth, tileheight, 3, 1, false
+          );
+          tiledata = transposeArray2d(tiledata, tilewidth*tileheight, 3);
+          break;
+
+        case ImageTileEncoding.JPG:
+        case ImageTileEncoding.JPEG:
+          /* Fragile heuristic with 10-minute read into the standard */
```

Comment:
We definitely don't want to have to parse the jpeg/png headers here to determine the sizes.  Instead, we should modify the `decodeJpeg` and `decodePng` async computation interfaces to return the actual dimensions, and allow it to be unspecified.  There could also be an option to skip the jpeg transpose if that is helpful.  That will presumably require some minor changes to existing callsites.

Reply:
Related: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray/@@species says `TypedArray`s can be subclassed, so it may be viable to stick `width`/`height` on `Uint8Array`, and existing code will ignore/slice. Assuming this really survives a `postMessage()`.

Reply:
unfortunately `postMessage` will not transfer extra properties like that on `TypedArray`

Reply:
I can add the other set of changes, just there may be a better idea for naming what's `BasicBitmap` above, and its `uint8Array` field (`width` and `height` could stay, I guess).